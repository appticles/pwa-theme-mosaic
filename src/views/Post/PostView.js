import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Container, Image, Header, Label, Icon, Modal, Transition } from 'semantic-ui-react';
import Moment from 'react-moment';
import PropTypes from 'prop-types';
import _ from 'lodash';

import { postPropType } from '../PostsCarousel/reducer';
import './style.css';

export default class PostView extends Component {
  constructor(props) {
    super(props);
    this.getImage = this.getImage.bind(this);
    this.handleOpen = this.handleOpen.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.state = {
      modalOpen: false,
      visible: true,
    };
  }

  getImage(sourceImg) {
    let imageSource;
    if (sourceImg) {
      imageSource = sourceImg[0].source_url;
      return imageSource;
    }
    imageSource = 'https://placeholdit.co//i/555x650';
    return imageSource;
  }

  handleOpen() {
    this.setState({ modalOpen: true, visible: !this.state.visible });
  }

  handleClose() {
    this.setState({ modalOpen: false, visible: !this.state.visible });
  }

  render() {
    const { post } = this.props;
    const { author } = post._embedded;
    const categoriesList = post._embedded['wp:term'];
    const featuredMedia = post._embedded['wp:featuredmedia'];
    // set path routes
    let goBack = {};
    let path = {};
    if (_.isNil(this.props.category.categorySlug)) {
      goBack = '/';
      path = '/post/' + post.slug + '/' + post.id + '/comments';
    } else {
      goBack = '/category/' + this.props.category.categorySlug + '/' + this.props.category.categoryId;
      path =
        '/category/' + this.props.category.categorySlug + '/' + this.props.category.categoryId + '/post/' + post.slug + '/' + post.id + '/comments';
    }

    return (
      <Container className="post">
        <Link to={goBack}>
          <Icon size="big" name="chevron left" />
        </Link>
        <Image src={this.getImage(featuredMedia)} />
        <Container textAlign="justified">
          {categoriesList[0].map(category => (
            <Label color="teal" key={category.name}>
              {category.name}
            </Label>
          ))}
          <Header>
            <div dangerouslySetInnerHTML={{ __html: post.title.rendered }} />
          </Header>
          <Header.Subheader>
            &nbsp;by&nbsp;<b>{author[0].name}</b>,&nbsp;<Moment format="MMMM DD, YYYY">{post.date}</Moment>
          </Header.Subheader>
          <div dangerouslySetInnerHTML={{ __html: post.content.rendered }} />
        </Container>
        <Modal
          open={this.state.modalOpen}
          onClick={this.handleClose}
          trigger={
            <Transition.Group animation="fade up" duration="500" className="transition-container">
              {this.state.visible && <Icon circular name="share alternate" size="large" onClick={this.handleOpen} />}
            </Transition.Group>
          }
          basic
        >
          <Modal.Actions>
            <Link to={path}>
              <Icon name="comment" size="large" circular inverted color="grey" />
            </Link>
            <a href={'https://m.facebook.com/sharer.php?u=' + post.link}>
              <Icon name="facebook f" size="large" circular inverted color="blue" />
            </a>
            <a href={'https://twitter.com/intent/tweet?text=' + encodeURIComponent(post.title.rendered) + ' ' + post.link}>
              <Icon name="twitter" size="large" circular inverted color="teal" />
            </a>
            <a href={'https://plus.google.com/share?url=' + post.link}>
              <Icon name="google plus" size="large" circular inverted color="red" />
            </a>
          </Modal.Actions>
        </Modal>
      </Container>
    );
  }
}
PostView.propTypes = {
  post: postPropType.isRequired,
  category: PropTypes.shape({
    categorySlug: PropTypes.string.isRequired,
    categoryId: PropTypes.string.isRequired,
  }).isRequired,
};
