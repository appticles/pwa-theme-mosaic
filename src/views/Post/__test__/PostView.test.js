import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import PostView from '../PostView';

describe('Post card', () => {
  it('should render post with image placeholder', () => {
    const mockPost = {
      id: 1,
      title: {
        rendered: 'Mock title',
      },
      author: 45,
      _embedded: {
        author: 'Some mock author',
        'wp:term': [[{ id: 1, name: 'Some category' }, { id: 2, name: 'Some category' }, { id: 3, name: 'Some category' }]],
      },
      categories: [2, 54, 23, 6],
      slug: 'mock-slug',
      content: {
        rendered: 'Some content that should be rendered',
        protected: false,
      },
      excerpt: {
        rendered: 'Some excerpt to be rendered',
        protected: true,
      },
      date: 'someMockDate', //TO DO : date validation => class Date
      // TO DO : proptype for image src
    };
    const category = {
      match: {
        params: {
          categoryId: '32',
          categorySlug: 'some-slug',
        },
      },
    };
    const output = shallow(<PostView post={mockPost} category={category} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render post with featured image', () => {
    const mockPost = {
      id: 1,
      title: {
        rendered: 'Mock title',
      },
      author: 45,
      _embedded: {
        author: 'Some mock author',
        'wp:term': [[{ id: 1, name: 'Some category' }, { id: 2, name: 'Some category' }, { id: 3, name: 'Some category' }]],
        'wp:featuredmedia': 'some-post-image',
      },
      categories: [2, 54, 23, 6],
      slug: 'mock-slug',
      content: {
        rendered: 'Some content that should be rendered',
        protected: false,
      },
      excerpt: {
        rendered: 'Some excerpt to be rendered',
        protected: true,
      },
      date: 'someMockDate', //TO DO : date validation => class Date
      // TO DO : proptype for image src
    };
    const category = {
      match: {
        params: {
          categoryId: '32',
          categorySlug: 'some-slug',
        },
      },
    };
    const output = shallow(<PostView post={mockPost} category={category} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
