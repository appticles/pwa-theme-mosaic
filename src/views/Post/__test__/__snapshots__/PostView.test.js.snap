// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post card should render post with featured image 1`] = `
<Container
  className="post"
>
  <Icon
    as="i"
    name="chevron left"
    onClick={[Function]}
    size="big"
  />
  <Image
    as="img"
    ui={true}
  />
  <Container
    textAlign="justified"
  >
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Header>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "Mock title",
          }
        }
      />
    </Header>
    <HeaderSubheader>
       by 
      <b />
      , 
      <t
        ago={false}
        calendar={false}
        element="time"
        filter={[Function]}
        format="MMMM DD, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      >
        someMockDate
      </t>
    </HeaderSubheader>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "Some content that should be rendered",
        }
      }
    />
  </Container>
  <Modal
    basic={true}
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    onClick={[Function]}
    open={false}
    trigger={
      <TransitionGroup
        animation="fade up"
        className="transition-container"
        duration="500"
      >
        <Icon
          as="i"
          circular={true}
          name="share alternate"
          onClick={[Function]}
          size="large"
        />
      </TransitionGroup>
    }
  >
    <Icon
      as="i"
      inverted={true}
      name="close"
      onClick={[Function]}
      size="large"
    />
    <ModalActions>
      <Link
        replace={false}
        to="/post/mock-slug/1/comments"
      >
        <Icon
          as="i"
          circular={true}
          color="grey"
          inverted={true}
          name="comment"
          size="large"
        />
      </Link>
      <a
        href="https://m.facebook.com/sharer.php?u=undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="blue"
          inverted={true}
          name="facebook f"
          size="large"
        />
      </a>
      <a
        href="https://twitter.com/intent/tweet?text=Mock%20title undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="teal"
          inverted={true}
          name="twitter"
          size="large"
        />
      </a>
      <a
        href="https://plus.google.com/share?url=undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="red"
          inverted={true}
          name="google plus"
          size="large"
        />
      </a>
    </ModalActions>
  </Modal>
</Container>
`;

exports[`Post card should render post with image placeholder 1`] = `
<Container
  className="post"
>
  <Icon
    as="i"
    name="chevron left"
    onClick={[Function]}
    size="big"
  />
  <Image
    as="img"
    src="https://placeholdit.co//i/555x650"
    ui={true}
  />
  <Container
    textAlign="justified"
  >
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Label
      color="teal"
      key="Some category"
    >
      Some category
    </Label>
    <Header>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "Mock title",
          }
        }
      />
    </Header>
    <HeaderSubheader>
       by 
      <b />
      , 
      <t
        ago={false}
        calendar={false}
        element="time"
        filter={[Function]}
        format="MMMM DD, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      >
        someMockDate
      </t>
    </HeaderSubheader>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "Some content that should be rendered",
        }
      }
    />
  </Container>
  <Modal
    basic={true}
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    onClick={[Function]}
    open={false}
    trigger={
      <TransitionGroup
        animation="fade up"
        className="transition-container"
        duration="500"
      >
        <Icon
          as="i"
          circular={true}
          name="share alternate"
          onClick={[Function]}
          size="large"
        />
      </TransitionGroup>
    }
  >
    <Icon
      as="i"
      inverted={true}
      name="close"
      onClick={[Function]}
      size="large"
    />
    <ModalActions>
      <Link
        replace={false}
        to="/post/mock-slug/1/comments"
      >
        <Icon
          as="i"
          circular={true}
          color="grey"
          inverted={true}
          name="comment"
          size="large"
        />
      </Link>
      <a
        href="https://m.facebook.com/sharer.php?u=undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="blue"
          inverted={true}
          name="facebook f"
          size="large"
        />
      </a>
      <a
        href="https://twitter.com/intent/tweet?text=Mock%20title undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="teal"
          inverted={true}
          name="twitter"
          size="large"
        />
      </a>
      <a
        href="https://plus.google.com/share?url=undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="red"
          inverted={true}
          name="google plus"
          size="large"
        />
      </a>
    </ModalActions>
  </Modal>
</Container>
`;
